OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,286 @@
+        void setValueForKey(const std::string& key,
+                            const ValueType& value) {
+            ValueArray<ValueType> value_array{};
+            value_array.upd().push_back(SimTK::Value<ValueType>{value});
+            _dictionary.setValueArrayForKey(key, value_array);
+        }
+
+        void removeValueForKey(const std::string& key) {
+            _dictionary.removeValueArrayForKey(key);
+        }
+
+        std::vector<std::string> getKeys() const {
+            return _dictionary.getKeys();
+        }
+
+    private:
+        ValueArrayDictionary _dictionary;
+    }; // ValueDictionary
+
+public:
+    using TableMetaData       = ValueDictionary;
+    using DependentsMetaData  = ValueArrayDictionary;
+    using IndependentMetaData = ValueDictionary;
+
+    AbstractDataTable()                                      = default;
+    AbstractDataTable(const AbstractDataTable&)              = default;
+    AbstractDataTable(AbstractDataTable&&)                   = default;
+    AbstractDataTable& operator=(const AbstractDataTable&)   = default;
+    AbstractDataTable& operator=(AbstractDataTable&&)        = default;
+    virtual std::unique_ptr<AbstractDataTable> clone() const = 0;
+    virtual ~AbstractDataTable()                             = default;
+
+    /** Get metadata associated with the table.                               */
+    const TableMetaData& getTableMetaData() const {
+        return _tableMetaData;
+    }
+
+    /** Update metadata associated with the table.                            */
+    TableMetaData& updTableMetaData() {
+        return _tableMetaData;
+    }
+
+    /** Get number of rows.                                                   */
+    size_t getNumRows() const {
+        return computeNumRows();
+    }
+
+    /** Get number of columns.                                                */
+    size_t getNumColumns() const {
+        return computeNumColumns();
+    }
+    
+    /** Get metadata associated with the independent column.                  */
+    const IndependentMetaData& getIndependentMetaData() const {
+        return _independentMetaData;
+    }
+    
+    /** Set metadata associated with the independent column.                  */
+    void 
+    setIndependentMetaData(const IndependentMetaData& independentMetaData) {
+        _independentMetaData = independentMetaData;
+        validateIndependentMetaData();
+    }
+
+    /** Get metadata associated with the dependent columns.                   */
+    const DependentsMetaData& getDependentsMetaData() const {
+        return _dependentsMetaData;
+    }
+    
+    /** Set metadata associated with the dependent columns.                   */
+    void 
+    setDependentsMetaData(const DependentsMetaData& dependentsMetaData) {
+        _dependentsMetaData = dependentsMetaData;
+        validateDependentsMetaData();
+    }
+
+protected:
+    virtual size_t computeNumRows() const            = 0;
```

Comment:
> It's true that in some cases we tell users to call Super::extendX() in their implementation, but we never expect them to call Super::X().

OK, thanks, that's nicer. But in any case `extend` should imply "adding on to" rather than "replacing". It seems awkward if sometimes it means "invoke Super" and other times it doesn't. I don't think a special verb is needed in the "replace" case since that is the typical use of virtual methods and is always the behavior of pure virtuals. I do like the "non virtual interface" design pattern though, which Shrinidhi used here where the virtuals are not part of the public API. With that pattern you often need two similar names, one for the public and one for the protected API. I tend just to stick "Virtual" at the end of the protected one; I've used "Impl" also but I don't like it as much.